src/Lean/Meta/ExprTraverse.lean

```diff
@@ -0,0 +1,96 @@
+/-
+Copyright (c) 2022 E.W.Ayers. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: E.W.Ayers
+-/
+import Lean.Meta.Basic
+import Lean.SubExpr
+
+namespace Lean.Meta
+
+open Lean.SubExpr (Pos)
+open Lean.SubExpr.Pos
+
+variable {M} [Monad M] [MonadLiftT MetaM M] [MonadControlT MetaM M]
+
+/-- Convert a traversal function to a form without the `Pos` argument. -/
+private def forgetPos (t : (Pos → Expr → M Expr) → (Pos → Expr → M Expr)) (visit : Expr → M Expr) (e : Expr) : M Expr :=
+  t (fun _ => visit) Pos.root e
+
+/-- Similar to `traverseLambda` but with an additional pos argument to track position. -/
+def traverseLambdaWithPos
+  (f : Pos → Expr → M Expr) (p : Pos) (e : Expr) : M Expr := visit #[] p e
```

Comment:
Alternative formulation: make

```
structure ExprTraverseContext with
  (pos : Pos)
  (fvars : Array Expr)
```

and then

```
traverseXXXWithPos (f : Expr → ReaderT ExprTraverseContext Expr) : Expr → ReaderT ExprTraverseContext Expr
```

Or perhaps instead include the variable instances `[MonadReader ExprTraverseContext M]` and `[MonadReaderWith ExprTraverseContext M]`?